### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by StephenDsouza90**

### CODE REVIEW

File: part_b/part_b.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	fmt.Scan(&startValue)
+
+	fmt.Println("Enter end value: ")
+	fmt.Scan(&endValue)
+
+	start := time.Now()
+
+	allSteps := getNumberOfStepsForEachM(startValue, endValue)
+
+	duration := time.Since(start)
```
Comment (pp-dsouzajude) [inline part_b/part_b.go]: Why a line space?

File: part_b/part_b.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+	fmt.Printf("\nAll steps for each number within a range of numbers from %v to %v \n\n", startValue, endValue)
+	fmt.Println(allSteps)
+
+	fmt.Printf("\n--- Took %v seconds ---\n\n", duration.Seconds())
+
```
Comment (pp-dsouzajude) [inline part_b/part_b.go]: Why so many line spaces?

File: part_b/part_b.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	// The number and its corresponding steps are stored in a dict,
+	// where key is number num and steps to compute the number is the value.
+	for num := startValue; num <= endValue; num++ {
+		steps := utils.GetNumberOfSteps(num)
+		allSteps[num] = steps
+	}
+
+	// allSteps : Dict with m as the key and it's corresponding steps as the value
+	return allSteps
+
```
Comment (pp-dsouzajude) [inline part_b/part_b.go]: No need for a line space.

File: utils/numberofstepsforeachm.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+				countSteps++
+			} else {
+				m = (m*3 + 1) / 2
+				countSteps += 2
+			}
+
+		}
+
+		allSteps[num] = countSteps
+
```
Comment (pp-dsouzajude) [inline utils/numberofstepsforeachm.go]: No need for a line space.

File: utils/numberofstepsforeachm.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			//	it can further be divided by 2 in the same step hence skipping one loop iteration
+			//	but still maintaining the steps.
+			if m%2 == 0 {
+				m = m / 2
+				countSteps++
+			} else {
+				m = (m*3 + 1) / 2
+				countSteps += 2
+			}
+
```
Comment (pp-dsouzajude) [inline utils/numberofstepsforeachm.go]: No need for a line space.

File: part_c/part_c.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+	fmt.Scan(&startValue)
+
+	fmt.Println("Enter end value: ")
+	fmt.Scan(&endValue)
+
+	start := time.Now()
+
+	// The func OptimizedGetNumberOfStepsForEachM() uses an optimized algorithm for computation.
+	allSteps := utils.OptimizedGetNumberOfStepsForEachM(startValue, endValue)
+
```
Comment (pp-dsouzajude) [inline part_c/part_c.go]: No need for a line space

File: part_c/part_c.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+	duration := time.Since(start)
+
+	fmt.Printf("\nAll steps for each number within a range of numbers from %v to %v \n\n", startValue, endValue)
+	fmt.Println(allSteps)
+
```
Comment (pp-dsouzajude) [inline part_c/part_c.go]: No need for a line space

File: part_c/part_c.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+	fmt.Printf("\n--- Took %v seconds ---\n\n", duration.Seconds())
+
```
Comment (pp-dsouzajude) [inline part_c/part_c.go]: No need for a line space

File: part_b/part_b.go (hunk 1 above)
Comment (StephenDsouza90) [reply part_b/part_b.go]: Removed line

File: part_b/part_b.go (hunk 2 above)
Comment (StephenDsouza90) [reply part_b/part_b.go]: Removed line space

File: part_b/part_b.go (hunk 3 above)
Comment (StephenDsouza90) [reply part_b/part_b.go]: Removed line space

File: utils/numberofstepsforeachm.go (hunk 4 above)
Comment (StephenDsouza90) [reply utils/numberofstepsforeachm.go]: Removed line space

File: utils/numberofstepsforeachm.go (hunk 5 above)
Comment (StephenDsouza90) [reply utils/numberofstepsforeachm.go]: Removed line space

File: part_c/part_c.go (hunk 6 above)
Comment (StephenDsouza90) [reply part_c/part_c.go]: Removed line space

File: part_c/part_c.go (hunk 7 above)
Comment (StephenDsouza90) [reply part_c/part_c.go]: Removed line space

File: part_c/part_c.go (hunk 8 above)
Comment (StephenDsouza90) [reply part_c/part_c.go]: Removed line space

File: utils/numberofstepsforeachm.go [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package utils
+
+import "math"
+
+// OptimizedGetNumberOfStepsForEachM : This function optimizes the Collatz Conjecture for a given range of numbers
+//	between startValue and endValue and returns the number of steps to reach 1.
```
Comment (pp-dsouzajude) [inline utils/numberofstepsforeachm.go]:
```suggestion
// between startValue and endValue and returns the number of steps to reach 1.
```

File: utils/numberofstepsforeachm.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+func OptimizedGetNumberOfStepsForEachM(startValue, endValue int) map[int]int {
+	//	startValue : starting number of the range
+	//	endValue : ending number of the range
```
Comment (pp-dsouzajude) [inline utils/numberofstepsforeachm.go]: This is part of the function definition, so try explaining what these variables are above the function.

File: utils/numberofstepsforeachm.go [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+	allSteps := make(map[int]int)
+
+	for num := startValue; num <= endValue; num++ {
+		countSteps := 0
+
+		m := num
+		for m != 1 {
+			//	Optimization 1:
```
Comment (pp-dsouzajude) [inline utils/numberofstepsforeachm.go]: Why so much space b/w `//` and `Optimization 1:`?

```suggestion
			// Optimization 1:
```

Comment (pp-dsouzajude) [reply utils/numberofstepsforeachm.go]: Also fix it for the rest.

File: part_b/part_b.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	fmt.Println(allSteps)
+	fmt.Printf("\n--- Took %v seconds ---\n\n", duration.Seconds())
+}
+
+// getNumberOfStepsForEachM : For each number num within a range of numbers between startValue and endValue,
+// this function returns the number of steps.
+func getNumberOfStepsForEachM(startValue, endValue int) map[int]int {
+	// startValue : starting number of the range
+	// endValue : ending number of the range
+
```
Comment (pp-dsouzajude) [inline part_b/part_b.go]: This comment is a duplication. It's already explained in the function definition.

File: utils/numberofsteps.go [hunk 13]
```diff
@@ -0,0 +1,24 @@
+package utils
+
+// GetNumberOfSteps : A function that counts the number of steps it takes for any given positive number to reach 1.
+func GetNumberOfSteps(userInput int) int {
+	// userInput : any given positive number from N
```
Comment (pp-dsouzajude) [inline utils/numberofsteps.go]: Move this comment to the function definition and remove it from here.

Comment (StephenDsouza90) [reply utils/numberofsteps.go]: Removed comment as it is not necessary and easily understood.

File: part_b/part_b.go (hunk 12 above)
Comment (StephenDsouza90) [reply part_b/part_b.go]: Removed comment as it is not necessary and easily understood.

File: utils/numberofstepsforeachm.go (hunk 10 above)
Comment (StephenDsouza90) [reply utils/numberofstepsforeachm.go]: Removed comment as it is not necessary and easily understood.

**Review by pp-dsouzajude — APPROVED** [review]
LGTM

**Pull request merged**